### PR TITLE
Make product list filter counts dependent on product visibility configuration

### PR DESCRIPTION
**Pull request by allanpaiste**
Fixes the issue that was caused by two factors: 
1. Visibility attribute was not indexed in EAV
2. Visibility filter was never applied due to a in_array check that compared int and string values which resulted in false-negative check.

Honestly tried to add a tests to this change as well, but instantiating product collection is a MAJOR head-ache due to things getting created, accessed and type-checked left and right - so - after seeing that just the setup for a test was going to be 100+ lines long - decided against creating it as it would just be unmaintainable.

Addresses issue: https://github.com/magento/magento2/issues/5670

### CODE REVIEW

File: app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php [hunk 1]
```diff
@@ -265,7 +265,7 @@ public function addFieldToFilter($field, $condition = null)
 
         $this->getSearchCriteriaBuilder();
         $this->getFilterBuilder();
-        if (!is_array($condition) || !in_array(key($condition), ['from', 'to'])) {
+        if (!is_array($condition) || !in_array((string)key($condition), ['from', 'to'])) {
```
Comment (v-r) [inline app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php]: we can use strict flag: like in_array(key($condition), ['from', 'to'], true)

Comment (allanpaiste) [inline app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php]: true that. was not aware of the 3rd argument :)

**Comment by tkacheva** [comment]
@allanpaiste please look into issues with your PR

File: app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php (hunk 1 above)
Comment (mombrea) [reply app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php]: I can confirm that the suggested change by @v-r along with this pull request do solve the problem. We need this fix which we've made on our local install. @allanpaiste any chance you can update this one line and submit the PR again? I could make a new request but I'd just be copying your work!

Comment (allanpaiste) [reply app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php]: changed & pushed. I don't mind if my work gets copied as long as Magento gets better :)

Comment (allanpaiste) [reply app/code/Magento/CatalogSearch/Model/ResourceModel/Fulltext/Collection.php]: Okay it seems that the build fails with pretty cryptic error. Considerign that develop branch itself is all in red, I would rather wait the PR target to go to be back in green before looking into that one. 

Considering the narrow scope of the change I really doubt that it's caused by this branch.

**Comment by shuaiZend** [comment]
@allanpaiste  i used 2.1.2,when i replace AbstractEav.php from https://github.com/magento/magento2/pull/6035/files

and flush cache, and reindex..

```
`[root@199 magento2]# bin/magento index:reindex 
Design Config Grid index has been rebuilt successfully in 00:00:02
Customer Grid index has been rebuilt successfully in 00:00:02
Category Products index has been rebuilt successfully in 00:00:00
Product Categories index has been rebuilt successfully in 00:00:00
Product Price index has been rebuilt successfully in 00:00:00
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '254-99-1-4' for key 'PRIMARY', query was: INSERT INTO `catalog_product_index_eav_idx` SELECT DISTINCT  `pid`.`entity_id`, `pid`.`attribute_id`, `pid`.`store_id`, IFNULL(pis.value, pid.value) AS `value` FROM (SELECT DISTINCT  `s`.`store_id`, `s`.`website_id`, `dd`.`attribute_id`, COALESCE(ds.value, dd.value) AS `value`, `cpe`.`entity_id` FROM `store` AS `s`
 LEFT JOIN `catalog_product_entity_int` AS `dd` ON dd.store_id = 0
 LEFT JOIN `catalog_product_entity_int` AS `ds` ON ds.store_id = s.store_id AND ds.attribute_id = dd.attribute_id AND ds.entity_id = dd.entity_id
 LEFT JOIN `catalog_product_entity_int` AS `d2d` ON d2d.store_id = 0 AND d2d.entity_id = dd.entity_id AND d2d.attribute_id = 97
 LEFT JOIN `catalog_product_entity_int` AS `d2s` ON d2s.store_id = s.store_id AND d2s.attribute_id = d2d.attribute_id AND d2s.entity_id = d2d.entity_id
 LEFT JOIN `catalog_product_entity` AS `cpe` ON cpe.entity_id = dd.entity_id WHERE (s.store_id != 0) AND ((ds.value IS NOT NULL OR dd.value IS NOT NULL)) AND (COALESCE(d2s.value, d2d.value) = 1)) AS `pid`
 LEFT JOIN `catalog_product_entity_int` AS `pis` ON pis.entity_id = pid.entity_id AND pis.attribute_id = pid.attribute_id AND pis.store_id = pid.store_id WHERE (pid.attribute_id IN('99')) AND (IFNULL(pis.value, pid.value) IS NOT NULL) AND (NOT(pis.value IS NULL AND pis.value_id IS NOT NULL))
Stock index has been rebuilt successfully in 00:00:00
Catalog Rule Product index has been rebuilt successfully in 00:00:00
Catalog Product Rule index has been rebuilt successfully in 00:00:00
Catalog Search index has been rebuilt successfully in 00:00:03`
```

how to fix it?

**Comment by maghamed** [comment]
Hi @allanpaiste 

How does current fix could address issue [#5670](https://github.com/magento/magento2/issues/5670), if Layered Navigation uses CatalogSearch Fulltext index and there are dedicated filters for visibility there specified in search_request.xml file
`<filter xsi:type="termFilter" name="visibility_filter" field="visibility" value="$visibility$"/>`
?

File: app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php [hunk 2]
```diff
@@ -148,6 +148,21 @@ public function reindexAttribute($attributeId, $isIndexable = true)
...
 
     /**
+     * Prepare product visibility attribute value index
+     *
+     * @param array $entityIds      the entity ids limitation
+     * @return $this
+     */
+    protected function _prepareVisibilityIndex($entityIds = null)
+    {
+        $attribute = $this->_eavConfig->getAttribute(\Magento\Catalog\Model\Product::ENTITY, 'visibility');
```
Comment (maghamed) [inline app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php:158]: It's better to use constant from Public Service Contract API 
\Magento\Catalog\Api\Data\ProductAttributeInterface::ENTITY_TYPE_CODE

File: app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php [hunk 3]
```diff
@@ -148,6 +148,21 @@ public function reindexAttribute($attributeId, $isIndexable = true)
...
+        $this->_prepareIndex($entityIds, $attribute->getAttributeId());
```
Comment (maghamed) [inline app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php:159]: why do we have additional call for _prepareIndex? if we had one before in chaining call.

This code looks awkward for me, as we have:

$this->_prepareIndex()->_prepareRelationIndex()->_prepareIndex($entityIds, $attribute->getAttributeId())->_removeNotVisibleEntityFromIndex();

The fact that we should to call the same method twice says about code smell

**Review by maghamed — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php (hunk 3 above)
Comment (mombrea) [reply app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php:159]: The additional call to _prepareIndex is necessary for this to work. Removing it results in the same incorrect visibility behavior.

**Pull request closed without merging**